### PR TITLE
Change m_bytes to unsigned in FixedBytesType

### LINKED ISSUE #3965: Change m_bytes to unsigned in FixedBytesType
The member variable `m_bytes` is signed, although it can never be negative. We should change its type to be unsigned.

### PR DESCRIPTION
**Pull request by djl-gg**
Fixes #3965.

@axic @chriseth 

Could I get dibs on this merge request since I started work on it initially and commented on the issue? Was pretty excited to contribute until I saw another PR open.

### CODE REVIEW

**Comment by chriseth** [comment]
You first requested to work on the issue, so no worries. Please change the type of the member variable `m_bytes`, not only the constructor parameter.

**Comment by djl-gg** [comment]
This should be good for another review whenever yall have time.

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -1281,9 +1281,9 @@ shared_ptr<FixedBytesType> FixedBytesType::smallestTypeForLiteral(string const&
 	return shared_ptr<FixedBytesType>();
 }
 
-FixedBytesType::FixedBytesType(int _bytes): m_bytes(_bytes)
+FixedBytesType::FixedBytesType(unsigned int _bytes): m_bytes(_bytes)
 {
-	solAssert(m_bytes >= 0 && m_bytes <= 32,
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Please remove the `>=0` check instead of changing semantics.

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -1281,9 +1281,9 @@ shared_ptr<FixedBytesType> FixedBytesType::smallestTypeForLiteral(string const&
 	return shared_ptr<FixedBytesType>();
 }
 
-FixedBytesType::FixedBytesType(int _bytes): m_bytes(_bytes)
+FixedBytesType::FixedBytesType(unsigned int _bytes): m_bytes(_bytes)
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: We usually just use `unsigned` instead of `unsigned int`.

File: libsolidity/ast/Types.h [hunk 3]
```diff
@@ -527,7 +527,7 @@ class FixedBytesType: public Type
 	int numBytes() const { return m_bytes; }
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: This should also return `unsigned`.

Comment (chriseth) [reply libsolidity/ast/Types.h]: And then of course, at every place this is used, it should be changed to `unsigned`. At some point, if the changes go to far, then of course we can convert to int again, but let's change a bit more for now.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/ast/Types.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/ast/Types.cpp]: Shouldn't there be an assert for `m_bytes != 0` though? In what case can it be 0?

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: There was a case in the past where zero length bytes was possible, even if not as a written type itself.

Comment (axic) [reply libsolidity/ast/Types.cpp]: The parser enforces `0 < m <= 32`, probably good idea to leave the assert as is then.

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: I think `""` can have a type of `bytes0` (of course only when using `var`).

My point is: This is about a refactoring, it should not change semantics.

Comment (axic) [reply libsolidity/ast/Types.cpp]: Lets strengthen that in a separate PR then, but it doesn't seem to be the case:
```
var a = "";
bytes1 b = a;
```

```
Error: Type string memory is not implicitly convertible to expected type bytes1.
```

File: libsolidity/ast/Types.h (hunk 3 above)
Comment (axic) [reply libsolidity/ast/Types.h]: And in many cases those places have assertions for negative numbers which will not be needed after this change.

**Comment by axic** [comment]
@danieljoonlee please rebase, the `!= 0` fix has been merged.

**Comment by axic** [comment]
Next time please do not close and open a new PR, rather do a rebase: `git rebase origin/develop`. Also it makes sense to work on a branch and not on `develop` directly to avoid confusion.

**Pull request closed without merging**